package/MDAnalysis/analysis/base.py

```diff
@@ -125,14 +125,105 @@
 import inspect
 import logging
 import itertools
+import warnings
 
 import numpy as np
 from MDAnalysis import coordinates
 from MDAnalysis.core.groups import AtomGroup
 from MDAnalysis.lib.log import ProgressBar
+from typing import Callable, Iterable
+from importlib import import_module
+from functools import partial
 
 logger = logging.getLogger(__name__)
 
+class ParallelExecutor(object):
+    """Class that executes parallel computations for AnalysisBase (but potentially for other purposes)
+    using one of the available backends"""
+
+    @classmethod
+    @property
+    def possible_backends(cls):
+        return ("local", "dask", "multiprocessing", "dask.distributed")
+
+    def __init__(self, backend: str, n_workers: int, client: object = None):
+        # validate 'client' argument
+        if client is not None and backend is not None:
+            raise ValueError(
+                (f"If 'client' is not None, backend is deduced and must be None, but here "
+                f'{backend=}, {client=}')
+            )
+        if client is not None:
+            # assuming it's a dask distributed client
+            from dask.distributed import Client as DistributedClient
+            if not isinstance(client, DistributedClient):
+                raise ValueError(f"client should be dask.distributed.Client instance, got {type(client)}")
+        self.client = client
+
+        # validate 'backend' argument
+        if backend is not None:
+            if backend not in ParallelExecutor.possible_backends:
+                raise ValueError(
+                    f"Invalid {backend=}: should be one of {ParallelExecutor.possible_backends}"
+                )
+            if backend == "local":
+                if n_workers > 1:
+                    raise ValueError("Can be only 1 worker with backend='local', got {n_workers=")
+            else:
+                try:
+                    import_module(backend)
+                except ImportError:
+                    raise ValueError(
+                        "Please install {backend} module "
+                        "with 'python3 -m pip install {backend} "
+                        "to run this")
+        self.backend = backend
+
+        # validate 'n_workers' argument
+        if n_workers and n_workers < 0:
+            raise ValueError(f"n_workers should be non-negative, got {n_workers=}")
+        self.n_workers = n_workers
+
+    def apply(self, func: Callable, computations: list) -> list:
```

Comment:
This function looks very general. Should it be lifted out of the class (and take a `backend` as input)?

Free functions are easier to test and to re-use.

Reply:
I don't think that it would make much sense -- this whole class was the point of having this function, but I believe that backend should be configured first -- for this purpose, I have a class around it.

Reply:
If the whole class was the point of having this function, shouldn't the class be a functor? The function seems very general to me and potentially useful, so lifting it out of the class (and, eventually, in a separate module), together with the associated functions called within would make it easily available if needed, but more importantly easily testable.

If you are interested, there is a great talk about this topic (in C++ not Python, but the underlying concepts apply to OOP vs free functions generally): [Free Your Functions!](https://www.youtube.com/watch?v=WLDT1lDOsb4).

Reply:
I know about the functors, but I believe it's relatively unnecessary here -- if we were to use this class in at least 2 different contexts, perhaps it'd be better to lift it up. But for now I believe we're using it exactly here and nowhere else, hence I don't see any points in introducing additional complexity with the class location and functionality.

Reply:
My point is that the complexity is not added but reduced by lifting these general functions out. But if you are happier with this solution it's fine by me!